src/Terminal/ui/TerminalInput.tsx

```diff
@@ -44,6 +44,11 @@ export function TerminalInput(): React.ReactElement {
   const [value, setValue] = useState(command);
   const [postUpdateValue, setPostUpdateValue] = useState<{ postUpdate: () => void } | null>();
   const [possibilities, setPossibilities] = useState<string[]>([]);
+  const [searchTerm, setSearchTerm] = useState<string>("");
+  const [searchResultsIndex, setSearchResultsIndex] = useState<number>(-1);
+  const [searchResults, setSearchResults] = useState<string[]>([]);
+  const [currentSearchPossibility, setCurrentSearchPossibility] = useState<string>("");
+  const [autofilledValue, setAutofilledValue] = useState<boolean>(false);
```

Comment:
To me, this seems like a lot of unnecessary values are stored...
What I would do is something like this:
- Remove searchTerm, just use value instead
- Remove currentSearchPossibility, use the first value of searchResults as the current value
- Remove searchResultsIndex, shift the searchResults instead
- Make searchResults be a computed value

The idea is to make all of the information you need be the array. The first value is the currently proposed command. Pressing up again shifts the array left, pushing the value from the front to the back.

Reply:
@G4mingJon4s Be aware that shifting has a complexity of O(n), compared to indexing being O(1)

Reply:
Yeah, that is why I originally held an index instead. In addition it seems like updating the index triggers a re-render, unlike shifting an array, so having to allocate a new array each time seems expensive, I think I'll re-implement the index instead, after doing this, unless there is a good reason not to?